File: RL-VizLib/src/rlVizLib/general/ParameterHolder.java
package rlVizLib.general;

import java.awt.Component;
import java.util.Map;
import java.util.StringTokenizer;
import java.util.TreeMap;
import java.util.Vector;

public class ParameterHolder {

	public static final int intParam=0;
	public static final int doubleParam=1;
	public static final int boolParam=2;
	public static final int stringParam=3;

	//This a straight port of my bt-glue C++ Code, might simplify in Java

	Map<String, Integer> intParams=new TreeMap<String, Integer>();
	Map<String, Double> doubleParams=new TreeMap<String, Double>();
	Map<String, String> stringParams=new TreeMap<String, String>();
	Map<String, Boolean> boolParams=new TreeMap<String, Boolean>();

	Map<String, Integer> allParams=new TreeMap<String, Integer>();
	Map<String, String> aliases=new TreeMap<String, String>();


	Vector<Integer> allParamTypes=new Vector<Integer>();
	Vector<String> allParamNames=new Vector<String>();
	Vector<String> allAliases=new Vector<String>();


	public ParameterHolder(){

	}
	

	public boolean isNull(){
		return (allParams.size()==0);
	}

	public ParameterHolder(final String theString){
		this();
		StringTokenizer iss=new StringTokenizer(theString,"_");

		int numParams;
		String thisParamName;

		int thisParamType;

		//Make sure the first bit of this isn't NULL!
		if(iss.nextToken().equals("NULL"))
			return;

		
		numParams=Integer.parseInt(iss.nextToken());

		for(int i=0;i<numParams;i++){
			thisParamName=iss.nextToken();
			thisParamType=Integer.parseInt(iss.nextToken());

			if(thisParamType==intParam){
				int thisParamValue=Integer.parseInt(iss.nextToken());
				addIntegerParam(thisParamName,thisParamValue);
			}
			if(thisParamType==doubleParam){
				double thisParamValue=Double.parseDouble(iss.nextToken());
				addDoubleParam(thisParamName,thisParamValue);
			}
			if(thisParamType==boolParam){
				boolean thisParamValue=Boolean.parseBoolean(iss.nextToken());
				addBooleanParam(thisParamName,thisParamValue);
			}
			if(thisParamType==stringParam){
				addStringParam(thisParamName,iss.nextToken());
			}
		}

		//Alias time
		int numAliases;
		numAliases=Integer.parseInt(iss.nextToken());


		for(int i=0;i<numAliases;i++){
			String thisAlias=iss.nextToken();
			String thisTarget=iss.nextToken();
			setAlias(thisAlias,thisTarget);
		}


	}



	public void setAlias(String thisAlias, String thisTarget) {
		if(thisAlias.contains(":")||thisAlias.contains("_")){
			System.err.println("The name or alias of a parameter cannot contain a space or : or _");
			Thread.dumpStack();
			System.exit(1);
		}
		
		if(allParams.get(thisTarget)==null){
			System.err.println("Careful, you are setting an alias of: "+thisAlias+" to original: "+thisTarget+" but the original isn't in the parameter set!");
			Thread.dumpStack();
			System.exit(1);
		}
		aliases.put(thisAlias,thisTarget);
		allAliases.add(thisAlias);
	}


	public void addStringParam(String thisParamName, String thisParamValue) {
		if(thisParamName.contains(":")||thisParamName.contains("_")){
			System.err.println("The ParameterName " + thisParamName + " with Parameter Value " + thisParamValue +  " cannot contain a : or _");
			Thread.dumpStack();
			System.exit(1);
		}
		
		if(thisParamValue.contains(":")||thisParamValue.contains("_")){
			System.out.println("The ParameterName " + thisParamName + " with ParameterValue " + thisParamValue + " cannot contain a : or _");
			Thread.dumpStack();
			System.exit(1);
		}
			
		addStringParam(thisParamName);
		setStringParam(thisParamName,thisParamValue);
	}



	public void setStringParam(String thisParamAlias, String thisParamValue) {
		String name=getAlias(thisParamAlias);
		if(!allParams.containsKey(name)){
			System.err.println("Careful, you are setting the value of parameter: "+name+" but the parameter hasn't been added...");
		}
		
		if(thisParamValue.contains(":")||thisParamValue.contains("_")){
			System.out.println("The ParameterValue " + thisParamValue + " cannot contain a : or _");
			Thread.dumpStack();
			System.exit(1);
		}
		stringParams.put(name, thisParamValue);
		
	}



	private void genericNewParam(String thisParamName){
		if(thisParamName.contains(":")||thisParamName.contains("_")){
			System.out.println("The ParameterName " + thisParamName + " cannot contain a : or _");
			Thread.dumpStack();
			System.exit(1);
		}
		allParamNames.add(thisParamName);
		setAlias(thisParamName,thisParamName);
	}
	public void addStringParam(String thisParamName) {
		if(thisParamName.contains(":")||thisParamName.contains("_")){
			System.out.println("The ParameterName " + thisParamName + " cannot contain a : or _");
			Thread.dumpStack();
			System.exit(1);
		}
		allParams.put(thisParamName,stringParam);
		allParamTypes.add(stringParam);
		genericNewParam(thisParamName);
	}

	public void addIntegerParam(String thisParamName) {
		if(thisParamName.contains(":")||thisParamName.contains("_")){
			System.out.println("The ParameterName " + thisParamName + " cannot contain a : or _");
			Thread.dumpStack();
			System.exit(1);
		}
		allParams.put(thisParamName,intParam);
		allParamTypes.add(intParam);
		genericNewParam(thisParamName);
	}

	public void addDoubleParam(String thisParamName) {
		if(thisParamName.contains(":")||thisParamName.contains("_")){
			System.out.println("The ParameterName " + thisParamName + " cannot contain a : or _");
			Thread.dumpStack();
			System.exit(1);
		}
		allParams.put(thisParamName,doubleParam);
		allParamTypes.add(doubleParam);
		genericNewParam(thisParamName);
	}


	public void addDoubleParam(String thisParamName, double thisParamValue) {
		if(thisParamName.contains(":")||thisParamName.contains("_")){
			System.out.println("The ParameterName " + thisParamName + " cannot contain a : or _");
			Thread.dumpStack();
			System.exit(1);
		}
		addDoubleParam(thisParamName);
		setDoubleParam(thisParamName,thisParamValue);
	}


	public void setDoubleParam(String thisParamAlias, double thisParamValue) {
		if(thisParamAlias.contains(":")||thisParamAlias.contains("_")){
			System.out.println("The ParameterAlias " + thisParamAlias + " cannot contain a : or _");
			Thread.dumpStack();
			System.exit(1);
		}
		String name=getAlias(thisParamAlias);
		if(!allParams.containsKey(name)){
			System.err.println("Careful, you are setting the value of parameter: "+name+" but the parameter hasn't been added...");
		}
		doubleParams.put(name, thisParamValue);
	}


	public void addBooleanParam(String thisParamName) {
		
		if(thisParamName.contains(":")||thisParamName.contains("_")){
			System.out.println("The ParameterName " + thisParamName + " cannot contain a : or _");
			Thread.dumpStack();
			System.exit(1);
		}
		allParams.put(thisParamName,boolParam);
		allParamTypes.add(boolParam);
		genericNewParam(thisParamName);
	}


	public void addBooleanParam(String thisParamName, boolean thisParamValue) {
		if(thisParamName.contains(":")||thisParamName.contains("_")){
			System.out.println("The ParameterName " + thisParamName + " cannot contain a : or _");
			Thread.dumpStack();
			System.exit(1);
		}
		addBooleanParam(thisParamName);
		setBooleanParam(thisParamName,thisParamValue);
	}


	public void setBooleanParam(String thisParamAlias, boolean thisParamValue) {
		if(thisParamAlias.contains(":")||thisParamAlias.contains("_")){
			System.out.println("The ParameterAlias " + thisParamAlias + " cannot contain a : or _");
			Thread.dumpStack();
			System.exit(1);
		}
		String name=getAlias(thisParamAlias);
		if(!allParams.containsKey(name)){
			System.err.println("Careful, you are setting the value of parameter: "+name+" but the parameter hasn't been added...");
		}
		boolParams.put(name, thisParamValue);
	}
	
	public void addIntegerParam(String thisParamName, int thisParamValue) {
		if(thisParamName.contains(":")||thisParamName.contains("_")){
			System.out.println("The ParameterName " + thisParamName + " cannot contain a : or _");
			Thread.dumpStack();
			System.exit(1);
		}
		addIntegerParam(thisParamName);
		setIntegerParam(thisParamName,thisParamValue);
	}



	public void setIntegerParam(String thisParamAlias,int thisParamValue) {
		if(thisParamAlias.contains(":")||thisParamAlias.contains("_")){
			System.out.println("The ParameterAlias " + thisParamAlias + " cannot contain a : or _");
			Thread.dumpStack();
			System.exit(1);
		}
		String name=getAlias(thisParamAlias);
		if(!allParams.containsKey(name)){
			System.err.println("Careful, you are setting the value of parameter: "+name+" but the parameter hasn't been added...");
		}
		intParams.put(name, thisParamValue);
	}



	private String getAlias(String thisParamAlias) {
		if(thisParamAlias.contains(":")||thisParamAlias.contains("_")){
			System.out.println("The ParameterAlias " + thisParamAlias + " cannot contain a  : or _");
			Thread.dumpStack();
			System.exit(1);
		}
		if(!aliases.containsKey(thisParamAlias)){
			System.err.println("You wanted to look up original for alias: "+thisParamAlias+", but that alias hasn't been set");
			Thread.dumpStack();
			System.exit(-1);
		}
		return  aliases.get(thisParamAlias);

	}






	public	String stringSerialize() {
		StringBuffer outs=new StringBuffer();

		//Do this here instead of externally later when we're ready
		outs.append("PARAMHOLDER_");
		//First, write the number of param names
		outs.append(allParamNames.size());
		outs.append("_");

		for(int i=0;i<allParamNames.size();i++){
			outs.append(allParamNames.get(i));
			outs.append("_");

			int paramType=allParamTypes.get(i);

			outs.append(paramType);
			outs.append("_");

			if(paramType==intParam)outs.append(getIntegerParam(allParamNames.get(i)));
			if(paramType==doubleParam)outs.append(getDoubleParam(allParamNames.get(i)));
			if(paramType==boolParam)outs.append(getBooleanParam(allParamNames.get(i)));
			if(paramType==stringParam)outs.append(getStringParam(allParamNames.get(i)));
			outs.append("_");
		}

		//Now write all of the aliases
		outs.append(allAliases.size());
		outs.append("_");

		for(int i=0;i<allAliases.size();i++){
			outs.append(allAliases.get(i));
			outs.append("_");
			outs.append(getAlias(allAliases.get(i)));
			outs.append("_");
		}

		return outs.toString();
	}



	public String getStringParam(String theAlias) {
		//Convert from an alias to the real name
		String name=getAlias(theAlias);

		if(!allParams.containsKey(name)){System.out.println("Careful, you are getting the value of parameter: "+name+" but the parameter hasn't been added...");System.exit(1);}
		if(!stringParams.containsKey(name)){System.out.println("Careful, you are getting the value of parameter: "+name+" but the parameter isn't a String parameter...");System.exit(1);}

		return stringParams.get(name);
	}



	public double getDoubleParam(String theAlias) {
		//Convert from an alias to the real name
		String name=getAlias(theAlias);

		if(!allParams.containsKey(name)){System.out.println("Careful, you are getting the value of parameter: "+name+" but the parameter hasn't been added...");System.exit(1);}
		if(!doubleParams.containsKey(name)){System.out.println("Careful, you are getting the value of parameter: "+name+" but the parameter isn't a double parameter...");System.exit(1);}

		return doubleParams.get(name);
	}

	public boolean getBooleanParam(String theAlias) {
		//Convert from an alias to the real name
		String name=getAlias(theAlias);

		if(!allParams.containsKey(name)){System.out.println("Careful, you are getting the value of parameter: "+name+" but the parameter hasn't been added...");System.exit(1);}
		if(!boolParams.containsKey(name)){System.out.println("Careful, you are getting the value of parameter: "+name+" but the parameter isn't a bool parameter...");System.exit(1);}

		return boolParams.get(name);
	}


	public int getIntegerParam(String theAlias) {
		//Convert from an alias to the real name
		String name=getAlias(theAlias);

		if(!allParams.containsKey(name)){System.out.println("Careful, you are getting the value of parameter: "+name+" but the parameter hasn't been added...");System.exit(1);}
		if(!intParams.containsKey(name)){System.out.println("Careful, you are getting the value of parameter: "+name+" but the parameter isn't an int parameter...");System.exit(1);}

		return intParams.get(name);
	}

public String toString(){
	StringBuffer theBuffer=new StringBuffer();
	for(int i=0;i<getParamCount();i++){
		int thisParamType=getParamType(i);
		String thisParamName=getParamName(i);

		switch (thisParamType) {
		case ParameterHolder.boolParam:
			theBuffer.append("boolParam: "+thisParamName+" = "+getBooleanParam(thisParamName));
			break;
		case ParameterHolder.intParam:
			theBuffer.append("intParam: "+thisParamName+" = "+getIntegerParam(thisParamName));
			break;
		case ParameterHolder.doubleParam:
			theBuffer.append("doubleParam: "+thisParamName+" = "+getDoubleParam(thisParamName));
			break;
		case ParameterHolder.stringParam:
			theBuffer.append("stringParam: "+thisParamName+" = "+getStringParam(thisParamName));
			break;
		}
		theBuffer.append("\n");
	}
	return theBuffer.toString();
}
	
//	int ParameterHolder::getParamCount(){
//	return allParamNames.size();
//	}
//	String ParameterHolder::getParamName(int which){
//	return allParamNames[which];
//	}
//	PHTypes ParameterHolder::getParamType(int which){
//	return allParamTypes[which];
//	}

//	bool ParameterHolder::supportsParam(String alias){
//	return (aliases.count(alias)!=0);
//	}

	public static ParameterHolder makeTestParameterHolder(){
		ParameterHolder p= new ParameterHolder();
		
		p.addDoubleParam("Alpha",.1);
		p.addDoubleParam("epsilon",.03);

		p.addIntegerParam("StepCount",5);
		p.addIntegerParam("Tiles",16);

		p.addStringParam("AgentName","Dave");
		p.addStringParam("AgentOccupation","Winner");

		p.addBooleanParam("ISCool",false);
		p.addBooleanParam("ISFast",true);

		return p;
	}
	public static void main(String []args){

		ParameterHolder p=makeTestParameterHolder();
		String serializedVersion = p.stringSerialize();
		System.out.println("serialized: "+serializedVersion);
		
		ParameterHolder unpackedP=new ParameterHolder(serializedVersion);

		System.out.println(p.getDoubleParam("Alpha"));
		System.out.println(p.getDoubleParam("epsilon"));
		System.out.println(p.getIntegerParam("StepCount"));
		System.out.println(p.getIntegerParam("Tiles"));
		System.out.println(p.getStringParam("AgentName"));
		System.out.println(p.getStringParam("AgentOccupation"));
		System.out.println(p.getBooleanParam("ISCool"));
		System.out.println(p.getBooleanParam("ISFast"));
		
		System.out.println("---");
		
		System.out.println(unpackedP.getDoubleParam("Alpha"));
		System.out.println(unpackedP.getDoubleParam("epsilon"));
		System.out.println(unpackedP.getIntegerParam("StepCount"));
		System.out.println(unpackedP.getIntegerParam("Tiles"));
		System.out.println(unpackedP.getStringParam("AgentName"));
		System.out.println(unpackedP.getStringParam("AgentOccupation"));
		System.out.println(unpackedP.getBooleanParam("ISCool"));
		System.out.println(unpackedP.getBooleanParam("ISFast"));

}

	public int getParamCount() {
		return allParams.size();
	}

	public int getParamType(int i) {
		return allParamTypes.get(i);
	}

	public String getParamName(int i) {
		return allParamNames.get(i);
	}


}

File: rl-competition/includes/ParameterHolder.py
# Parameter Holder for Passing Parameters Around
# Copyright (C) 2007, Brian Tanner brian@tannerpages.com (http://brian.tannerpages.com/)
# 
# This program is free software; you can redistribute it and/or
# modify it under the terms of the GNU General Public License
# as published by the Free Software Foundation; either version 2
# of the License, or (at your option) any later version.
# 
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
# 
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.

INT_PARAM=0
DOUBLE_PARAM=1
BOOL_PARAM=2
STRING_PARAM=3

class ParameterHolder:
	intParams = {}
	doubleParams = {}
	boolParams = {}
	stringParams = {}
	
	allParams = {}
	#we'll let everything be an alias to itself, and then we'll always just look up aliases
	aliases = {}

	allParamNames = []
	allParamTypes = []
	allAliases = []
	
	def __init__(self,stringRep=None):
		if stringRep == None:
			return
		arrayRep = stringRep.split('_')
		
		numParams = arrayRep.pop(0)
		numParams = int(numParams)
		for i in range(numParams):
			thisParamName = arrayRep.pop(0)
			thisParamType = int(arrayRep.pop(0))
			if thisParamType == INT_PARAM:
				iParamValue = int(arrayRep.pop(0))
				addIntegerParam(thisParamName,iParamValue)
			elif thisParamType == DOUBLE_PARAM:
				fParamValue = float(arrayRep.pop(0))
				addDoubleParam(thisParamName,fParamValue)
			elif thisParamType == BOOL_PARAM:
				bParamValue = arrayRep.pop(0)
				if bParamValue == "true":
					addBoolParam(thisParamName,True)
				else:
					addBoolParam(thisParamName,False)
			elif thisParamType == STRING_PARAM:
				sParamValue = arrayRep.pop(0)
				addStringParam(thisParamName,sParamValue)
		
		numAliases = arrayRep.pop(0)
		numAliases = int(numAliases)
		for i in range(numAliases):
			thisAlias = arrayRep.pop(0)
			thisTarget = arrayRep.pop(0)
			setAlias(thisAlias,thisTarget)


	def stringSerialize(self): #return string
		outString = "PARAMHOLDER_%d_" % (len(allParamNames))
		for i in range(len(allParamNames)):
			paramType = allParamTypes[i]
			outString = outString + "%s_%d_" % (allParamNames[i],paramType)
			if paramType == INT_PARAM:
				outString = outString + "%d_" % (getIntegerParam(allParamNames[i]))
			elif paramType == DOUBLE_PARAM:
				outString = outString + "%f_" % (getDoubleParam(allParamNames[i]))
			elif paramType == BOOL_PARAM:
				if (getBoolParam(allParamNames[i])):
					outString = outString + "true_"
				else
					outString = outString + "false_"
			else
				outString = outString + getStringParam(allParamNames[i])
			
		outString = outString + "%d_" % (len(allAliases))
		for i in range(len(allAliases)):
			outString = outString + "%s_%s_" % (allAliases[i],getAlias(allAliases[i]))
		
		return outString


	def getAlias(self,alias): #return string
		if not aliases.has_key(alias):
			sys.stderr.write("You wanted to look up original for alias: %s, but that alias hasn't been set\n" % (alias))
			sys.exit(-1)
		return aliases[alias]
	
	def supportsParam(self,alias): #return bool
		return aliases.has_key(alias)

	def setAlias(self,alias,original): #void
		if not allParams.has_key(original):
			sys.stderr.write("C++ Parameter Holder::Careful, you are setting an alias of:%s to: %s but: %s isn't in the parameter set\n" % (alias,original,original))
			sys.exit(1)
		aliases[alias] = original
		allAliases.append(alias)
		

	def setIntegerParam(self,alias,value): #void
		name = getAlias(alias)
		if not allParams.has_key(name):
			sys.stderr.write("Careful, you are setting the value of parameter: %s but the parameter hasn't been added...\n" % (name))
		intParams[name] = value
	
	def setDoubleParam(self,alias,value): #void
		name = getAlias(alias)
		if not allParams.has_key(name):
			sys.stderr.write("Careful, you are setting the value of parameter: %s but the parameter hasn't been added...\n" % (name))
		doubleParams[name] = value
	
	def setBoolParam(self,alias,value): #void
		name = getAlias(alias)
		if not allParams.has_key(name):
			sys.stderr.write("Careful, you are setting the value of parameter: %s but the parameter hasn't been added...\n" % (name))
		boolParams[name] = value
	
	def setStringParam(self,alias,value): #void
		name = getAlias(alias)
		if not allParams.has_key(name):
			sys.stderr.write("Careful, you are setting the value of parameter: %s but the parameter hasn't been added...\n" % (name))
		boolParams[name] = value

	def addIntegerParam(self,alias): #void
		allParams[name] = INT_PARAM
		allParamNames.append(name)
		allParamTypes.append(intParam)
		setAlias(name,name)
	
	def addDoubleParam(self,alias): #void
		allParams[name] = DOUBLE_PARAM
		allParamNames.append(name)
		allParamTypes.append(doubleParam)
		setAlias(name,name)
	
	def addBoolParam(self,alias): #void
		allParams[name] = BOOL_PARAM
		allParamNames.append(name)
		allParamTypes.append(boolParam)
		setAlias(name,name)
	
	def addStringParam(self,alias): #void
		allParams[name] = STRING_PARAM
		allParamNames.append(name)
		allParamTypes.append(stringParam)
		setAlias(name,name)

	#Should have done this a while ago
	def addIntegerParam(self,alias,defaultValue): #void
		addIntegerParam(name)
		setIntegerParam(name, defaultValue)
	
	def addDoubleParam(self,alias,defaultValue): #void
		addDoubleParam(name)
		setDoubleParam(name, defaultValue)
	
	def addBoolParam(self,alias,defaultValue): #void
		addBoolParam(name)
		setBoolParam(name, defaultValue)
	
	def addStringParam(self,alias,defaultValue): #void
		addStringParam(name)
		setStringParam(name, defaultValue)

	def getIntegerParam(self,alias): #int
		name = getAlias(alias)
		if not allParams.has_key(name):
			sys.stderr.write("Careful, you are getting the value of parameter: %s but the parameter hasn't been added...\n" % (name))
			sys.exit(1)
		if not intParams.has_key(name):
			sys.stderr.write("Careful, you are getting the value of parameter: %s but the parameter isn't an int parameter...\n" % (name))
			sys.exit(1)
		return intParams[name]
	
	def getDoubleParam(self,alias): #double
		name = getAlias(alias)
		if not allParams.has_key(name):
			sys.stderr.write("Careful, you are getting the value of parameter: %s but the parameter hasn't been added...\n" % (name))
			sys.exit(1)
		if not doubleParams.has_key(name):
			sys.stderr.write("Careful, you are getting the value of parameter: %s but the parameter isn't a double parameter...\n" % (name))
			sys.exit(1)
		return doubleParams[name]
	
	def getBoolParam(self,alias): #bool
		name = getAlias(alias)
		if not allParams.has_key(name):
			sys.stderr.write("Careful, you are getting the value of parameter: %s but the parameter hasn't been added...\n" % (name))
			sys.exit(1)
		if not boolParams.has_key(name):
			sys.stderr.write("Careful, you are getting the value of parameter: %s but the parameter isn't a boolean parameter...\n" % (name))
			sys.exit(1)
		return boolParams[name]
	
	def getStringParam(self,alias): #string
		name = getAlias(alias)
		if not allParams.has_key(name):
			sys.stderr.write("Careful, you are getting the value of parameter: %s but the parameter hasn't been added...\n" % (name))
			sys.exit(1)
		if not stringParams.has_key(name):
			sys.stderr.write("Careful, you are getting the value of parameter: %s but the parameter isn't a string parameter...\n" % (name))
			sys.exit(1)
		return stringParams[name]

	def getParamCount(self): #int
		return len(allParams)
	
	def getParamName(self,which): #string
		return allParamNames[which]
	
	def getParamType(self,which): #PHTypes
		return allParamTypes[which]



